crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+                    let ty_idx = u.int_in_range(0..=types.len() - 1)?;
+                    let ty = types[ty_idx].clone();
+                    defs.push(ModuleTypeDef::Import(crate::core::Import(
+                        module,
+                        field,
+                        crate::core::EntityType::Func(u32::try_from(ty_idx).unwrap(), ty),
+                    )));
+                }
+
+                // Export.
+                1 if !types.is_empty() => {
+                    let name = crate::unique_string(100, &mut exports, u)?;
+                    let ty_idx = u.int_in_range(0..=types.len() - 1)?;
+                    let ty = types[ty_idx].clone();
+                    defs.push(ModuleTypeDef::Export(
+                        name,
+                        crate::core::EntityType::Func(u32::try_from(ty_idx).unwrap(), ty),
+                    ));
+                }
+
+                // Type definition.
+                _ => {
+                    let ty = crate::core::arbitrary_func_type(u, &self.core_valtypes)?;
+                    types.push(ty.clone());
+                    defs.push(ModuleTypeDef::TypeDef(crate::core::Type::Func(ty)));
+                }
+            }
+            Ok(true)
+        })?;
+
+        Ok(ModuleType { defs })
+    }
+
+    fn with_types_scope<T>(&mut self, f: impl FnOnce(&mut Self) -> Result<T>) -> Result<T> {
+        self.types_scopes.push(vec![]);
+        let result = f(self);
+        self.types_scopes.pop();
+        result
+    }
+
+    fn current_type_scope(&self) -> &[Rc<Type>] {
+        self.types_scopes.last().unwrap()
+    }
+
+    fn current_type_scope_mut(&mut self) -> &mut Vec<Rc<Type>> {
+        self.types_scopes.last_mut().unwrap()
+    }
+
+    fn outer_types_scope(&self, count: u32) -> &[Rc<Type>] {
+        &self.types_scopes[self.types_scopes.len() - 1 - usize::try_from(count).unwrap()]
+    }
+
+    fn outer_type(&self, count: u32, i: u32) -> &Rc<Type> {
+        &self.outer_types_scope(count)[usize::try_from(i).unwrap()]
+    }
+
+    fn arbitrary_component_type(&mut self, u: &mut Unstructured<'a>) -> Result<ComponentType> {
+        let mut defs = vec![];
+        let mut imports = HashSet::new();
+        let mut exports = HashSet::new();
+
+        self.with_types_scope(|me| {
+            arbitrary_loop(u, 0, 100, |u| {
+                if !me.current_type_scope().is_empty() && u.arbitrary()? {
+                    // Imports.
+                    let name = crate::unique_string(100, &mut imports, u)?;
+                    let ty = u.int_in_range(0..=me.current_type_scope().len() - 1)?;
+                    let ty = u32::try_from(ty).unwrap();
+                    defs.push(ComponentTypeDef::Import(Import { name, ty }));
+                } else {
+                    // Instance type definitions.
+                    let def = me.arbitrary_instance_type_def(u, &mut exports)?;
+                    defs.push(ComponentTypeDef::Instance(def));
+                }
```

Comment:
My ignorance of the current spec is probably showing but I thought components were similar to modules where they have imports, exports, aliases, and possible type definitions themselves. That makes me naively think like things are missing here but wanted to confirm.

Reply:
The components have these things, but the component *types* don't, IIUC.